controller/network/path_test.go

```diff
@@ -302,3 +303,208 @@ func TestShortestPath(t *testing.T) {
+	ctx := db.NewTestContext(t)
+	defer ctx.Cleanup()
+
+	req := assert.New(t)
+
+	closeNotify := make(chan struct{})
+	defer close(closeNotify)
+
+	network, err := NewNetwork("test", nil, ctx.GetDb(), nil, NewVersionProviderTest(), closeNotify)
+	req.NoError(err)
+
+	addr := "tcp:0.0.0.0:0"
+	transportAddr, err := tcp.AddressParser{}.Parse(addr)
+	req.NoError(err)
+
+	r0 := newRouterForTest("r0", "", transportAddr, nil, 1, true)
+	network.Routers.markConnected(r0)
+
+	r1 := newRouterForTest("r1", "", transportAddr, nil, 2, false)
+	network.Routers.markConnected(r1)
+
+	r2 := newRouterForTest("r2", "", transportAddr, nil, 3, true)
+	network.Routers.markConnected(r2)
+
+	r3 := newRouterForTest("r3", "", transportAddr, nil, 4, true)
+	network.Routers.markConnected(r3)
+
+	link := newLink("l0")
+	link.SetStaticCost(2)
+	link.SetDstLatency(10 * 1_000_000)
+	link.SetSrcLatency(11 * 1_000_000)
+	link.Src = r0
+	link.Dst = r1
+	link.addState(newLinkState(Connected))
+	network.linkController.add(link)
+
+	link = newLink("l1")
+	link.SetStaticCost(5)
+	link.SetDstLatency(15 * 1_000_000)
+	link.SetSrcLatency(16 * 1_000_000)
+	link.Src = r0
+	link.Dst = r2
+	link.addState(newLinkState(Connected))
+	network.linkController.add(link)
+
+	link = newLink("l2")
+	link.SetStaticCost(9)
+	link.SetDstLatency(20 * 1_000_000)
+	link.SetSrcLatency(21 * 1_000_000)
+	link.Src = r1
+	link.Dst = r3
+	link.addState(newLinkState(Connected))
+	network.linkController.add(link)
+
+	link = newLink("l3")
+	link.SetStaticCost(13)
+	link.SetDstLatency(25 * 1_000_000)
+	link.SetSrcLatency(26 * 1_000_000)
+	link.Src = r2
+	link.Dst = r3
+	link.addState(newLinkState(Connected))
+	network.linkController.add(link)
+
+	path, cost, err := network.shortestPath(r0, r3)
+	req.NoError(err)
+	req.NotNil(t, path)
+	req.Equal(path[0], r0)
+	req.Equal(path[1], r2)
+	req.Equal(path[2], r3)
+
+	expected := 15 + 16 + 5 + 3 + // link1 cost and src and dest latency plus dest router cost
+		25 + 26 + 13 + 4 // link3 cost and src and dest latency plus dest router cost
+	req.Equal(int64(expected), cost)
+}
+
+func TestShortestPathWithOnlyUntraversableRouter(t *testing.T) {
+	ctx := db.NewTestContext(t)
+	defer ctx.Cleanup()
+
+	req := assert.New(t)
+
+	closeNotify := make(chan struct{})
+	defer close(closeNotify)
+
+	network, err := NewNetwork("test", nil, ctx.GetDb(), nil, NewVersionProviderTest(), closeNotify)
+	req.NoError(err)
+
+	addr := "tcp:0.0.0.0:0"
+	transportAddr, err := tcp.AddressParser{}.Parse(addr)
+	req.NoError(err)
+
+	r0 := newRouterForTest("r0", "", transportAddr, nil, 1, true)
+	network.Routers.markConnected(r0)
+
+	r1 := newRouterForTest("r1", "", transportAddr, nil, 2, false)
+	network.Routers.markConnected(r1)
+
+	link := newLink("l0")
+	link.SetStaticCost(2)
+	link.SetDstLatency(10 * 1_000_000)
+	link.SetSrcLatency(11 * 1_000_000)
+	link.Src = r0
+	link.Dst = r1
+	link.addState(newLinkState(Connected))
+	network.linkController.add(link)
+
```

Comment:
This test should pass. Even though r1 is not traversable, it's ok because the path is not traversing r1, it's terminating at r1.

Reply:
Ah I must have misunderstood the intent here. The way it was in my head is that data shouldn't traverse through that router. So we only care if the middle routes are nontraversable then?